### PR TITLE
Use `raw-dylib` in the std for non-x86 Windows

### PR DESCRIPTION
**Pull request by ChrisDenton**
The [`raw_dylib`](https://github.com/rust-lang/rust/issues/58713) feature was recently [stabilized for non-x86 Windows targets](https://github.com/rust-lang/rust/pull/99916). This is now in beta so std start to use it without feature flags. Eventually this may allowing linking Rust programs without needing import libraries.

I think the standard library is best placed to dog food before the beta becomes the stable the release. There hopefully shouldn't be any issues but if there are it'd be good to catch them early.

### CODE REVIEW

**Comment by ChrisDenton** [comment]
This PR does not use raw-dylib for kernel32 because the vcruntime needs it to be linked in some way. Also note that some ntdll we depend on are now linked using `raw-dylib`. The only reason these were dynamically loaded is that some old versions of the Windows SDK may not have the required import library. Using `raw-dylib` means it's no longer necessary.

**Comment by thomcc** [comment]
This is really nice. I kind of expect that this might hit some issues for some users use-cases, but I do think we ultimately want to do this, and this seems like the best way to root out such issues.

@bors r+

**Comment by ChrisDenton** [comment]
The error is:

```text
   = note: a.tcp_stress.8a6846bf-cgu.0.rcgu.o : error LNK2019: unresolved external symbol __imp_closesocket referenced in function _ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17he99d99762636b925E
```

Hm, so this is weird because it's only `closesocket` (from `ws2_32`) that fails with `raw-dylib`. What makes it special is that this is used in a small drop function that's `#[inline]` so it's being inlined into `__rust_begin_short_backtrace`: https://github.com/rust-lang/rust/blob/8b705839cd656d202e920efa8769cbe43a5ee269/library/std/src/os/windows/io/socket.rs#L202-L209

So the test framework loads the std dynamically but I guess `closesocket` is inlined statically. And unlike with normal links, it's not picking up that it also needs the import.

cc @dpaoliello

**Comment by ChrisDenton** [comment]
[Test successful](https://github.com/rust-lang/rust/actions/runs/3136264922/jobs/5092955027). Reverting test commit.

**Comment by dpaoliello** [comment]
> The error is:
> 
> ```
>    = note: a.tcp_stress.8a6846bf-cgu.0.rcgu.o : error LNK2019: unresolved external symbol __imp_closesocket referenced in function _ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17he99d99762636b925E
> ```
> 
> Hm, so this is weird because it's only `closesocket` (from `ws2_32`) that fails with `raw-dylib`. What makes it special is that this is used in a small drop function that's `#[inline]` so it's being inlined into `__rust_begin_short_backtrace`:
> 
> https://github.com/rust-lang/rust/blob/8b705839cd656d202e920efa8769cbe43a5ee269/library/std/src/os/windows/io/socket.rs#L202-L209
> 
> So the test framework loads the std dynamically but I guess `closesocket` is inlined statically. And unlike with normal links, it's not picking up that it also needs the import.
> 
> cc @dpaoliello

Interesting - can you please file and issue and assign it to me? I'm on vacation next week but will look into it the week after that.

I'm not sure if the best approach will be to have the declaring binary (in this case, std) include the symbol in its normal import library, or if the consuming binary should generate the symbol instead...

**Comment by thomcc** [comment]
I don't think this is waiting on review, so I'm punting it to the author.

That said, maybe S-blocked is more accurate...

@rustbot author

**Comment by ChrisDenton** [comment]
Issue created. This PR is blocked on #102714 so I'll set it to draft.

**Comment by thomcc** [comment]
This looks great! Thrilled to see that the underlying issue was fixable too!

@bors r+

**Comment by ChrisDenton** [comment]
Ok so the miri issue was that we're now importing `NtWriteFile` directly instead of loading dynamically so the shim had to be moved. Hopefully I did it right.

File: src/tools/miri/src/shims/windows/foreign_items.rs [hunk 1]
```diff
@@ -454,6 +454,68 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
                 this.write_null(dest)?;
             }
 
+            // I/O
```
Comment (RalfJung) [inline src/tools/miri/src/shims/windows/foreign_items.rs:457]:
```suggestion
```
This is still in the same 'incomplete shims' category as the above code.

**Review by RalfJung — COMMENTED** [review]
Looks good, just a comment nit!

File: library/std/src/sys/windows/c.rs [hunk 2]
```diff
@@ -1338,6 +1385,7 @@ compat_fn_with_fallback! {
     ) -> NTSTATUS {
         STATUS_NOT_IMPLEMENTED
     }
+    #[cfg(target_arch = "x86")]
     pub fn NtWriteFile(
```
Comment (RalfJung) [inline library/std/src/sys/windows/c.rs:1389]: Is still still using the dlsym on 32bit Windows? Then we'd have to keep that code in Miri.

Comment (ChrisDenton) [reply library/std/src/sys/windows/c.rs:1389]: Oh right! Yeah, we'll need to keep both for now.

Comment (ChrisDenton) [reply library/std/src/sys/windows/c.rs:1389]: Hm, should I simply duplicate the code or factor it out somewhere?

Comment (RalfJung) [reply library/std/src/sys/windows/c.rs:1389]: Please put it in a function somewhere... `windows/foreign_items.rs` seems like a reasonable place for now.

**Comment by RalfJung** [comment]
r=me on the Miri changes.

**Comment by ChrisDenton** [comment]
Ok, I'm closing this for now. It seems it would be wise to let the `raw_dylib` feature bake for longer before using it in std. There's a PR up which will use it in the compiler. Once that is merged and has been around for awhile I'll look to revisit this.

I'll keep this branch around though.

**Pull request closed without merging**